### DISCUSSION TITLE
Unable to bind on Accessors & Mutators that access casted Json

### DISCUSSION THREAD

**Discussion by roni-estein** [livewire/livewire#3140] [7. Bugs (archived)]
### Description
For the first time ever I'm stuck using a JSON mysql field. I can create an accessor simply enough however I can't find a way to save it into the DB, no matter what path i take I end up with an error that I can't work around.

### Exact steps to reproduce
I can make an RmMe for this if you like but either way you are stuck running migrations to see it unfold, please let me know if you want me to attach a series of file for that or if you would like to reproduce it locally on your own.

1. Have a model that has a json field.
2. cast the field to an array
3. create an accessor on the model
4. create a property setter on the model
5. add protected $appends = ['foo']

```

Class Thing extends Model{

 public function getFooAttribute()
    {
        return array_get($this->data, 'foo', 1);
    }
    
    public function setFooAttribute($value)
    {
        array_set($this->attributes,'data.foo',$value );
    }
```

In your components you'll need to wire:model="foo" on some component, mine is a select, and and it's wired to editing.foo, where editing contains my thing.

Everything work great until you need to save thing, I can't figure out how to save the the foo attribute unless it's bound to a DB column, and in this case, I can't change the DB.

### Stripped-down, copy-pastable code snippets

**Context**
- Livewire version: [e.g. 2.3.1]
- Laravel version: [e.g. 8.10.1=0]
- Browser: [e.g. Brave]

**Comment by roni-estein** [comment]
On a data table that closely follows the surge data tables setup, I had an element bound to a field stored in an attribute on a JSON property.

```
$rules = ['editing.attendees' => 'required|int:min,1',];
```

my blade field for this looked as follows:

```
<x-surge.input.group for="attendees" label="Number of People">
  <x-surge.input.select wire:model="editing.attendees" id="attendees" class="rounded">
    @foreach(range(1,8) as $number)
      <option value="{{ $number}}">{{ $number }} {{ str_plural('Person', $number) }}</option>
    @endforeach
  </x-surge.input.select>
</x-surge.input.group>
```

My model had the following class attributes

```
protected $casts = ['data'=>'json'];
protected $appends = ['attendees'];
```
as well as these accessors and mutators, in multiple forms, any variant possible you can think of in it's current state and with the casts and appends removed here are some samples:

```
    // so much pain these are just some of the myriad of attempts, 
    // nothing worked, using the data element as an array or 
    // direct access through attributes. All threw a myriad of errors
    // were called at unexpected times

    public function getAttendeesAttribute()
    {
//        try{

            $attribute = data_get($this->attributes, 'data','{}' );
            if(is_string($attribute)) {
                $data = json_decode($attribute);
                return $data->attendees ?? 1;

            }
//                dd('boom');
//            $attribute = $this->attributes['data'] ?? '{}';
////            Log::debug($attribute);
//
//            if(is_array($attribute))
//                return $attribute['attendees'];
//
//
//
//            }
//        }catch(\ErrorException $e){
//            dd($e);
//        }

    }
    
    public function setAttendeesAttribute($value)
    {
        dump('we are in set ... why? ' .$value);
        $json = data_get($this->attributes, 'data','{}' );
        if($json === '{}')
            $json = '{"attendees":"'.$value.'"}';
        else {
            $data = json_decode($json, true);

            array_set($this->attributes,'data', json_encode($data));
        }

    }
    
    // or these
    public function setAttendeesAttribute($value){
        data_set($this->attributes, 'data.attendees', $value);
    }
    
    public function getAttendeesAttribute(){
        return array_get($this->attributes, 'data.attendees', 1);
    }
```

However this removing all accessors and mutators and treating data as a simple array did work.

```
$rules = ['editing.data.attendees' => 'required|int:min,1',];

// and

<x-surge.input.group for="attendees" label="Number of People">
  <x-surge.input.select wire:model="editing.data.attendees" id="attendees" class="rounded">
    @foreach(range(1,8) as $number)
      <option value="{{ $number}}">{{ $number }} {{ str_plural('Person', $number) }}</option>
    @endforeach
  </x-surge.input.select>
</x-surge.input.group>
```

There is definitely something happing with setting when it comes to json attributes. It should be possible. 

@iksaku and I are looking further into this, can you let us know how we could best help to move this forward? We are checking to see if we can make a more complex laravel playground with models. Should be just make a project that can be cloned? I know it's not as good as simple element, but I'm not sure how we could distill this down.

Any advice welcome.

**Comment by squishythejellyfish** [comment]
👋 Oh Hi! I'm Squishy, the friendly jellyfish that manages Livewire issues.

I see this issue has been closed.

Here in the Livewire repo, we have an "issues can be closed guilt-free and without explanation" policy.

If for ANY reason you think this issue hasn't been resolved, PLEASE feel empowered to re-open it.

Re-opening actually helps us track which issues are a priority.

Reply "REOPEN" to this comment and we'll happily re-open it for you!

(More info on this philosophy here: https://twitter.com/calebporzio/status/1321864801295978497)

**Comment by iksaku** [comment]
REOPEN

**Comment by basepack** [comment]
I got the exact same with casting Spatie's DTO objects to JSON. I think livewire stumbles over the Modal casting part.
